src/index.ts

```diff
@@ -1,5 +1,33 @@
-export default function alwaysArray<T>(input: T | T[]): T[] {
-	return Array.isArray(input) ? input : [input];
+type SingleOrArray<T> = T | T[] | Iterable<T> | readonly T[];
+
+interface Options {
+	convertIterables: boolean;
+}
+
+/**
+ * Makes sure you're dealing with an array.
+ * @param input Input that will be converted to an array if it isn't already.
+ * @param options.convertIterables [false] Iterables will be spread into an array.
+ */
+export default function alwaysArray<T>(
+	input: SingleOrArray<T>,
+	options?: Options
+): T[] {
+	options = {
+		convertIterables: false,
+		...options
+	};
+
+	const isArray = Array.isArray(input);
+
+	const isIterable =
+		isArray ||
+		(options.convertIterables === true &&
+			input !== null &&
+			typeof input !== 'string' &&
+			typeof (input as Iterable<T>)[Symbol.iterator] === 'function');
+
+	return isIterable ? [...(input as Iterable<T>)] : [input as T];
```

Comment:
Having to assume Types for Typescript looks fishy to me. If its a simple module like this I would hope TypeScript is able to keep track of the types.

Maybe work with multiple return statements? For example getting rid of the simplest case:
```ts
if (Array.isArray(input)) {
  return […input]
}
```

TypeScript will now after this if input is not a T[] anymore so only T | Iterable<T> can be the case.

Problem with that: strings are Iterables too? Which is kinda messy then as the result of this isn't clear then: accepts iterators, string is iterable, what happens? The current code ignores string and just returns it as a string array then. But you have to know that from the code, you can't assume that from only looking at the input / output types.

Reply:
The idea with multiple return statements sounds pretty good!

```ts
if (Array.isArray(input)) {
	// Sanity check passes:
	// const x: T[] = input;
	return [...input]; // at this stage, 
}

if (
	options.convertIterables === true &&
	input !== null &&
	input instanceof Object &&
	typeof input[Symbol.iterator] === 'function'
) {
	// Sanity check passes:
	// const x: Iterable<T> = input;
	return [...input];
}

// Sanity check fails:
// const x: T = input;

// Error: Type 'SingleOrArray<T>' is not assignable to type 'T'.
return [input]; 
```

It works well too, except for the last statement. TypeScript still thinks `input` is anyone of the `SingleOrArray<T>` types...

The approach above is still flawed though, as an input of `new String('foo')` would be spread into an array as it's an instance of `Object`, so an `!(input instanceof String)` check would be needed too (which indicates to me that the solution is getting a bit too hairy).

> Problem with that: strings are Iterables too? Which is kinda messy then as the result of this isn't clear then: accepts iterators, string is iterable, what happens? The current code ignores string and just returns it as a string array then. But you have to know that from the code, you can't assume that from only looking at the input / output types.

Yeah, this bothered me too. What do you think about the fix using overloads in 693ec29?

Reply:
It's not possible to pass overloads from one method to another.

```ts
function foo(bar: string | string[]) {
  const arr = alwaysArray(bar) // fails as the correct overload can not be determined
}
```

Is it really needed to also accept Iterables? Seems like stuff like strings creates hiccups with them more often than it helps. If the user has a simple `T | T[]` input this `alwaysArray` is a shortcut. If the type of the user is more complex the user should care himself about this?

Personally I like using `T[]` when `T` might be an option as long as it's an internal method. Simpler interface and using `[value]` when calling is not that problematic. TypeScript will complain when there is something wrong. Spreaded arguments like `function foo(…bar: number[]` can also solve this. As I haven't used Iterables yet I'm not sure how they fit it.
Maybe two different methods might be more helpful for the user to decide what he actually needs:

```ts
function first(input: T | readonly T[]): T[];
function second(input: readonly T[] | Iterable<T>): T[];
```

In this case the string problem is solved:
`first('string')` will return `['string']` and `second('string')` will return `['s', 't', 'r', 'i', 'n', 'g']` as in both cases its clear which type string will be.

Reply:
> It's not possible to pass overloads from one method to another.

The example works fine for me :thinking: 

> Maybe two different methods might be more helpful

Good idea. e9c467c

> In this case the string problem is solved

I personally don't like spreading strings into an array, it doesn't seem sensible for the use-case. A note in the docs mentioning this behaviour is present: https://github.com/krmax44/always-array/commit/e9c467ce9acce20699926805bdb492a3f482015a#diff-04c6e90faac2675aa89e2176d2eec7d8R31